analysis/pseudobulk-bulk-prediction/scripts/prepare-ora-gene-sets.R

```diff
@@ -0,0 +1,132 @@
+# This script prepares gene sets for input to over-representation analysis (ORA) for a given ScPCA project as follows:
+# 1. Identify all unique consensus cell types across the project 
+# 2. Identify the PanglaoDB cell types that contributed to those consensus labels
+# 3. Using the scpca-nf PanglaoDB cell type reference for the given project, determine the genes in those PanglaoDB cell types
+# 4. These (unique) genes define the gene set for a given consensus cell type to test in ORA
+
+
+renv::load()
+library(optparse)
+
+
+# Parse options --------
+option_list <- list(
+  make_option(
+    "--celltype_dir",
+    type = "character",
+    help = "Directory with consensus cell types for the given project"
+  ),
+  make_option(
+    "--panglao_geneset_file",
+    type = "character",
+    help = "Path to gene set file used to cell type the given project"
+  ), 
+  make_option(
+    "--map_file",
+    type = "character",
+    help = "Path to TSV file mapping bulk sample and library ids, which is needed because not all samples are used in all projects"
+  ),
+  make_option(
+    "--output_file",
+    type = "character",
+    help = "Path to output TSV file to save gene sets."
+  ), 
+  make_option(
+    "--consensus_reference_url",
+    type = "character",
+    default = "https://raw.githubusercontent.com/AlexsLemonade/OpenScPCA-analysis/refs/heads/main/analyses/cell-type-consensus/references/consensus-cell-type-reference.tsv",
+    help = "URL with map between consensus cell types and PanglaoDB gene set labels"
+  ), 
+  make_option(
+    "--library_metadata_file",
+    type = "character",
+    default = here::here("s3_files", "scpca-library-metadata.tsv"),
+    help = "Path to ScPCA library metadata file."
+  )
+)
+opts <- parse_args(OptionParser(option_list = option_list))
+
+# Checks---------
+stopifnot(
+  "Panglao gene set file not found" = file.exists(opts$panglao_geneset_file), 
+  "Library metadata file not found" = file.exists(opts$library_metadata_file), 
+  "ID map file not found" = file.exists(opts$map_file)
+)
+
+# Prepare input files  -----------------
+consensus_ref_df <- readr::read_tsv(opts$consensus_reference_url)
+
+panglao_df <- readr::read_tsv(opts$panglao_geneset_file) |>
+  tidyr::pivot_longer(
+    -ensembl_id, 
+    names_to = "panglao_celltype", 
+    values_to = "gene_present"
+  )
+
+celltype_files <- list.files(
+  opts$celltype_dir, 
+  pattern = "_processed_consensus-cell-types\\.tsv\\.gz$", 
+  full.names = TRUE, 
+  recursive = TRUE
+) |>
+  purrr::set_names(
+    \(x) {stringr::str_split_i(basename(x), pattern = "_", i = 1)}
+  )
+
+# determine libraries to keep from the consensus cell types
+# we need to take this step because some projects don't use all their single-cell libraries
+keep_libraries <- readr::read_tsv(opts$map_file) |>
+  dplyr::select(scpca_sample_id) |>
```

Comment:
I would approach this differently. 
First, filter the library metadata to the project IDs and those that are in bulk and pull the sample IDs. This gives you the list of sample IDs that have bulk RNA-seq. 
Then filter the library metadata to seq unit in cell/ nucleus and sample ID in bulk list and pull the library IDs that way. 

This should eliminate needing that extra map file.

Reply:
For this, I'll also need to filter out multiplexed (which can be done by removing any rows where there is a `;` in the sample id), and add an argument for project id, but yeah this should be fine!